### PR TITLE
Improve promotion script and pipelines

### PR DESCRIPTION
**Pull request by ycedres**
In our efforts to improve and reduce complexity of our Salt release process, this PR makes the following improvements:

- Promotion script:
    -  Improve promotion script for it to be able to exclude packages and subprojects to copy.
    -  Allow execution of individual stages.
- Salt package promotion:
  - New parameter -> `recreate_mu_branches` to force MU branches recreation.
  - MU branches are now created by pipeline if they don't exist. If they already exist pipeline will fail by default unless using `recreate_mu_branches` parameter.
- Salt Bundle promotion:
  - Execute different stages separately.
- Cosmetic changes on displayed output.

### CODE REVIEW

**Comment by meaksh** [comment]
I've tested a modified version of the pipelines (pointing to my home project) and both pipelines seems to run fine as described:

https://ci.suse.de/view/Manager/view/salt/job/manager-salt-pipeline-test-pablo/
https://ci.suse.de/view/Manager/view/salt/job/manager-salt-bundle-pipeline-test-pablo/

I think this is ready for a review now.

Keep in mind, this is not yet prepared for making the execution of the pipeline to happen automatically (we still need to figure out the locking problem).

But, for now:
- Pipelines execution is separated in different stages.
- Promotion pipeline for classic Salt package is now using our "promote_package.py" script. No need to maintain package names inside pipelines anymore.
- Promotion can be triggered now without our intervention (as MU branches are created by the pipeline).

File: promote_packages.py [hunk 1]
```diff
@@ -11,90 +11,152 @@
...
     project_handler = Project(client)
     project_handler.set_config(prj, config=config)
 
+
 def has_link(client, project, package) -> bool:
     pkg_handler = Package(client)
-    pkg_files = pkg_handler.get_files(project,package)
-    linkinfo = etree.fromstring(etree.tostring(pkg_files).decode('utf-8')).find("linkinfo")
+    pkg_files = pkg_handler.get_files(project, package)
+    linkinfo = etree.fromstring(etree.tostring(pkg_files).decode("utf-8")).find(
```
Comment (agraul) [inline promote_packages.py:89]: Why is this "fromstring(tostring().decode("utf-8"))" needed?

**Review by agraul — COMMENTED** [review]
Looks good overall, I just have a few questions I left inline.

I also wonder how you want to manage the dependencies needed to run this. I'd prefer to keep them defined in the same place as the script, e.g. using one of the newer Python project management tools.

File: Jenkinsfile_promote_salt_packages [hunk 2]
```diff
@@ -9,49 +9,83 @@ properties([
...
 
-                echo "Check that 'products:testing' and 'products:testing:debian' are not set to MU branches"
-                sh "! curl https://build.opensuse.org/package/view_file/systemsmanagement:saltstack:products:testing/salt/_service?expand=1 | grep MU/${mu_version}"
-                sh "! curl https://build.opensuse.org/package/view_file/systemsmanagement:saltstack:products:testing:debian/salt/_service?expand=1 | grep MU/${mu_version}"
+                    dir('/tmp/salt-promote-pipeline-env') {
+                        if (salt_mu_branch != 0 || salt_packaging_mu_branch != 0) {
+                            echo "MU branches do not exist. Creating them"
+                            sh "git clone --branch openSUSE/release/${salt_version} --depth 1 https://github.com/openSUSE/salt"
+                            dir('/tmp/salt-promote-pipeline-env/salt') {
+                                sh "git checkout -b openSUSE/MU/${mu_version}"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:39]: Oh, nice. Thanks for the suggestion!

I'm used to use `git checkout -b` to create a new branch based on the current HEAD and switch to it. You mentioned `git-checkout` somehow does this differently, how is this?

Out of curiosity: is there any particular reason to use `switch --create` in favor of `checkout -b` besides of cosmetical reasons? As `git-switch` is a more explicit/concrete command while `checkout` is actually used for many different things.

I'm perfectly fine using `git-switch` instead, which as mentioned seems more concrete command, but I wonder if there is some more reasons. TIA!

File: Jenkinsfile_promote_salt_packages [hunk 3]
```diff
@@ -9,49 +9,83 @@ properties([
...
+                            echo "Successfully created and pushed 'MU/${mu_version}' branch to 'openSUSE/salt-packaging' repository"
+                        }
+                        else {
+                            if (params.recreate_salt_mu_branches) {
+                                // Forcing recreation of MU branches
+                                echo "The MU branches already exist, but pipeline was set to recreate them"
+                                sh "git clone --branch openSUSE/release/${salt_version} --depth 1 https://github.com/openSUSE/salt"
+                                dir('/tmp/salt-promote-pipeline-env/salt') {
+                                    sh "git checkout -b openSUSE/MU/${mu_version}"
+                                    sh "git push -f origin openSUSE/MU/${mu_version}"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:57]: Ok, sounds reasonable. Better to be on safe side in case of unexpected behavior.

File: Jenkinsfile_promote_salt_packages (hunk 2 above)
[the remark this answers is not in the document]
Comment (agraul) [reply Jenkinsfile_promote_salt_packages:39]: > I'm used to use `git checkout -b` to create a new branch based on the current HEAD and switch to it. You mentioned `git-checkout` somehow does this differently, how is this?
> 

I can't quite follow, can you rephrase that question? Are you confused because I used `git-checkout` instead of `git checkout`? That's just because many git commands are implemented in separated programs, e.g. `git foo` loads `git-foo`.

> Out of curiosity: is there any particular reason to use `switch --create` in favor of `checkout -b` besides of cosmetical reasons? As `git-switch` is a more explicit/concrete command while `checkout` is actually used for many different things.
> 

Not really, `git-switch` was only added to improve the UX with clearer commands (along with `git-restore`) in 2.23.0 ([Changelog](https://github.com/git/git/blob/master/Documentation/RelNotes/2.23.0.txt#L61))

File: Jenkinsfile_promote_salt_packages [hunk 4]
```diff
@@ -9,49 +9,83 @@ properties([
...
+                                    sh "git checkout -b MU/${mu_version}"
+                                    sh "git push -f origin MU/${mu_version}"
+                                }
+                                echo "Successfully recreated and pushed 'MU/${mu_version}' branch to 'openSUSE/salt-packaging' repository"
+                            }
+                            else {
+                                error("The MU branches were already created. Default promotion is not possible. Exiting.")
+                            }
+                        }
+                        sh "rm salt -rf && rm salt-packaging -rf"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:71]: In case of `error()`, that particular call to `rm` won't run but then, the general post execution cleanup will take care of removing the whole temporary environment (including those directories):

https://github.com/openSUSE/salt-package-promote-obs/blob/9700d9c981399bd4ca9880e8a1ba53ac1765c619/Jenkinsfile_promote_salt_packages#L125-L129

So, this shouldn't be an issue.

File: Jenkinsfile_promote_salt_packages [hunk 5]
```diff
@@ -9,49 +9,83 @@ properties([
...
+                    }
 
-                echo "Check the source tarball is properly named to salt_${salt_version}.orig.tar.gz in 'products:testing:debian'"
-                sh "curl https://build.opensuse.org/package/view_file/systemsmanagement:saltstack:products:testing:debian/salt/_service?expand=0 | grep salt_${salt_version}.orig.tar.g"
-            }
-        }
+                    echo "Check that 'products:testing' and 'products:testing:debian' are not set to MU branches"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:74]: This has to do with how we were doing for building the DEB packages, where pipeline was by then adjusting the branches configured at `_service`  files. Now, this is done differently, so the branch configured in `_service` file are not expected to change during the execution of the pipeline.

In `products:testing` and `products:testing:debian`, we have `_service` configured to point to our devel branch: `openSUSE/release/3006.0` (not really to an MU branch). This way, any time we push something new to `products:testing`, the services would run at `products:testing:debian` updating the tarball that is used to build DEB package.

The pipeline then expects that this is the case, so when `_service` is pointing to an MU branch, then this means something unexpected had happened and we don't really want to run the promotion at this point.

If the `_service` is NOT pointing to an MU branch, then pipeline assumes we are in a good shape, and will run the promotion.

We can keep this checks, which are harmless at the moment, and will makes us to notice in case the branches has changed.

File: Jenkinsfile_promote_salt_packages [hunk 6]
```diff
@@ -9,49 +9,83 @@ properties([
...
-                sh "osc copypac systemsmanagement:saltstack:products:testing python3-immutables systemsmanagement:saltstack:products"
-                sh "osc copypac systemsmanagement:saltstack:products:testing py26-compat-salt systemsmanagement:saltstack:products"
-                sh "osc copypac systemsmanagement:saltstack:products:testing py26-compat-tornado systemsmanagement:saltstack:products"
-                sh "osc copypac systemsmanagement:saltstack:products:testing py26-compat-msgpack-python systemsmanagement:saltstack:products"
-                sh "osc copypac systemsmanagement:saltstack:products:testing py27-compat-salt systemsmanagement:saltstack:products"
-                sh "osc copypac systemsmanagement:saltstack:products:old:testing salt systemsmanagement:saltstack:products:old"
+                sh "python3 promote_packages.py -s systemsmanagement:saltstack:products:testing -t systemsmanagement:saltstack:products packages"
+                sh "python3 promote_packages.py -s systemsmanagement:saltstack:products:old:testing -t systemsmanagement:saltstack:products:old packages"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:88]: Hmmm, good catch. Salt 3000 is EOL, we do not maintain them anymore.

I'm removing this part.

File: Jenkinsfile_promote_salt_packages [hunk 7]
```diff
@@ -68,14 +102,15 @@ pipeline {
...
-                        sh "osc copypac --keep-link systemsmanagement:saltstack:products:testing:debian salt systemsmanagement:saltstack:products:debian"
-                        sh "osc copypac systemsmanagement:saltstack:products:testing:debian python3-contextvars systemsmanagement:saltstack:products:debian"
-                        sh "osc copypac systemsmanagement:saltstack:products:testing:debian python3-immutables systemsmanagement:saltstack:products:debian"
-                        sh "osc copypac systemsmanagement:saltstack:products:testing:debian python3-zmq systemsmanagement:saltstack:products:debian"
-                        sh "osc copypac systemsmanagement:saltstack:products:testing:debian python3-gnupg systemsmanagement:saltstack:products:debian"
+                echo 'Promote "products:testing:debian" packages'
+                sh "osc copypac --keep-link systemsmanagement:saltstack:products:testing:debian salt systemsmanagement:saltstack:products:debian"
+                sh "python3 promote_packages.py -s systemsmanagement:saltstack:products:testing:debian -t systemsmanagement:saltstack:products:debian --exclude salt packages"
```
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages:110]: The way we have linked the "salt" package in our `:debian` subprojects is what makes us to do this workaround here from the normal way of promoting the rest of the packages.

We have to run `copypac --keep-link` only for the "salt" package in order to not break the structure and proper inheritance for "salt" package, but this is only required for "salt" package (not for the other dependencies).

So we exclude "salt" here as we do an explicit `copypac --keep-link` for `salt` package just in the line before, and then run the normal promotion for the rest of the packages.

File: Jenkinsfile_promote_salt_packages (hunk 2 above)
[the remark this answers is not in the document]
Comment (agraul) [reply Jenkinsfile_promote_salt_packages]: > [...] I was just wondering why you mentioned that:
> 
> _git-switch has a long option for "create branch and switch to it", while **git-checkout does not.**_
> 
> My understanding is that, here in this context, the behavior of both `git switch --create` and `git checkout -b` executions would be the same: create a new branch and swich to it.
> 
> So just wondered why you mentioned that `git checkout -b` wouldn't be creating new branch and switch to it.
> 

That's not what I meant. The emphasis should be on the **long option** part of the sentence. `git-checkout` only has a short option (`-b`) while `git-switch` has a short (`-c`) and a long (`--create`) option.

File: promote_packages.py (hunk 1 above)
Comment (meaksh) [reply promote_packages.py:89]: You are right. This is not really needed.

File: Jenkinsfile_promote_salt_packages (hunk 2 above)
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages]: Got it! Sorry for the misunderstanding!

File: Jenkinsfile_promote_salt_packages (hunk 6 above)
[the remark this answers is not in the document]
Comment (agraul) [reply Jenkinsfile_promote_salt_packages]: Those are not for Salt 3000, `products:old` is for SLE 11 / RHEL 6.

**Review by agraul — APPROVED** [review]
The changes look good, but the commit history could use a cleanup.

File: Jenkinsfile_promote_salt_packages (hunk 6 above)
[the remark this answers is not in the document]
Comment (meaksh) [reply Jenkinsfile_promote_salt_packages]: Oops, that is correct. I think it makes also sense to remove it from here. In case an update for Salt 2016.11.10 is really required, we will do "copypac" manually.
